crud/common/sharding_key.lua

```diff
@@ -115,6 +116,87 @@ function sharding_key_module.fetch_on_router(space_name)
     return cache.sharding_key_def
 end
 
+-- Make sure sharding key definition is a part of primary key.
+function sharding_key_module.is_part_of_pk(space_name, space_format,
+                                           primary_index, sharding_key_def)
+    dev_checks('string', 'table', 'table', 'table')
+
+    if cache.as_index_object[space_name] == nil then
+        cache.as_index_object[space_name] =
+            sharding_key_module.as_index_object(space_format, space_name,
+                                                sharding_key_def)
+    end
+    local sharding_key_as_index_obj = cache.as_index_object[space_name]
+
+    if cache.is_part_of_pk[space_name] ~= nil then
+        return cache.is_part_of_pk[space_name]
+    end
+
+    local is_part_of_pk = true
+    local pk_fieldno_map = utils.get_index_fieldno_map(primary_index.parts)
+    for _, part in ipairs(sharding_key_as_index_obj.parts) do
+        if pk_fieldno_map[part.fieldno] == nil then
+            is_part_of_pk = false
+            break
+        end
+    end
+    cache.is_part_of_pk[space_name] = is_part_of_pk
+
+    return is_part_of_pk
+end
+
+-- Build an array with sharding key values.
+function sharding_key_module.extract_from_index(sharding_key_def,
+                                                index_parts,
+                                                sharding_key_fieldno_map)
+    dev_checks('table', 'table', 'table')
+
+    local sharding_key = {}
+    for i, k in ipairs(sharding_key_def) do
+        local fieldno = index_parts[i].fieldno
+        if sharding_key_fieldno_map[fieldno] == true then
+            table.insert(sharding_key, k)
+        end
+    end
```

Comment:
Okay, if we assume that 'sharding_key_def' is 'pk' (value), we fill 'sharding_key' in the order of the primary key values. I mean, if the primary index is built on fields 1 and 2 and sharding key is defines as composition of fields 2 and 1, it seems, we're in the trouble. Is there a relevant test case? I would prefer a unit test, it is simpler.

Reply:
Good catch! I had only one unit test for this function and it's case didn't cover reverse order of fields in a sharding key. I have added more tests and fixed implementation.

```
    sharding_key.test_extract_from_index_sharding_key_direct_order ... (0.008s) Ok
    sharding_key.test_extract_from_index_sharding_key_reverse_order ... (0.007s) Ok
    sharding_key.test_extract_from_index_sharding_key_single_field ... (0.007s) Ok
    sharding_key.test_extract_from_index_sharding_key_none_fields ... (0.011s) Ok
```